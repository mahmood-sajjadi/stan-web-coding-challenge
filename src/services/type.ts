import type { ProgramType } from "../config";

export interface ListResponse {
    total:   number;
    entries: Entry<ProgramType>[];
}

export interface Entry<T extends ProgramType> {
    title:       string;
    description: string;
    programType: T;
    images:      Images;
    releaseYear: number;
}

export interface Images {
    "Poster Art": PosterArt;
}

export interface PosterArt {
    url:    string;
    width:  number;
    height: number;
}