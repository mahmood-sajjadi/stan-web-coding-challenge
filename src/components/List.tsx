import { useMemo } from "react";

type Props = {
    children?: React.ReactNode,
    style?: React.CSSProperties;
}

const listWrapperStyleBase = {
    display: 'grid',
    gridTemplateColumns: 'repeat(auto-fit, minmax(100px, 1fr))',
    gridGap: '5px',
  }

export function List(props: Props) {
    const { style, children } = props;

    const listWrapperStyle = useMemo(() => ({
        ...listWrapperStyleBase,
        ...style,
    }), [style]);

    return (<div style={listWrapperStyle}>{children}</div>);
}