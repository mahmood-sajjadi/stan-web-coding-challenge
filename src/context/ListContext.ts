import React from 'react';
import type { ListResponse } from '../services/type';

export type Status = 'pending' | 'success' | 'error'

export type ListContextType = {
    status: Status,
    error: string,
    result: ListResponse | undefined,
}

export const ListContext = React.createContext<ListContextType>({
    status: 'pending',
    error: '',
    result: undefined,
});