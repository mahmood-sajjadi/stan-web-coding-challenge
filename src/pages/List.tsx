import { useContext, useMemo } from "react";
import { getList } from "../helper/getList"
import { ListContext } from "../context/ListContext";
import type { ProgramType } from "../config";
import { Item } from "../components/Item";
import { List as ListWrapper } from "../components/List";
import { Loading } from "../components/Loading";

export function List(props: { programType: ProgramType }) {
    const { programType } = props;
    const list = useContext(ListContext);

    const listComponents = useMemo(() => {
        if (!list.result) {
            return null;
        }
        return getList(list.result, programType).map((entry) => (
            <Item key={entry.title}
                posterArtUrl={entry.images["Poster Art"].url}
                title={entry.title}
                className="glass"
            />
        ));
    }, [list.result, programType]);


    return (
        <Loading status={list.status}>
            <ListWrapper>
            {
                listComponents
            }
            </ListWrapper>
        </Loading>
    );
}