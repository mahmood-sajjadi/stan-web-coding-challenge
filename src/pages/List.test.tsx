import { render } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom';
import {List} from './List'
import { ListContext, type Status } from '../context/ListContext'
import type { ListResponse } from '../services/type'
import type { ProgramType } from '../config';

function testSetup(status: Status, error: string, result: ListResponse, programType: ProgramType) {
    
    return (
        <ListContext.Provider value={{
            status,
            error,
            result,
        }}>
          <MemoryRouter>
            <List programType={programType}/>
          </MemoryRouter>
        </ListContext.Provider>
    )
}

describe('List', () => {
  it('renders the List component', () => {
    const list = render(testSetup('success', '', {total: 0, entries: []}, 'movie'));
    expect(list).toMatchSnapshot();
  })

  it('renders the List component WITH values', () => {
    const list = render(testSetup('success', '', {total: 0, entries: [{
        description: 'test',
        images: {
            "Poster Art": {
                height: 1,
                url: '',
                width: 1,
            }
        },
        programType: 'movie',
        releaseYear: 1,
        title: 'abc'
    }]}, 'movie'));
    expect(list).toMatchSnapshot();
  })

  it('renders the List component WHEN error', () => {
    const list = render(testSetup('error', 'some error', {total: 0, entries: []}, 'movie'));
    expect(list).toMatchSnapshot();
  })

  it('renders the List component WHEN pending', () => {
    const list = render(testSetup('pending', '', {total: 0, entries: []}, 'movie'));
    expect(list).toMatchSnapshot();
  })
})