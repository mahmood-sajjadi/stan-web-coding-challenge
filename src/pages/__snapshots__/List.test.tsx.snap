// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`List > renders the List component 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="display: grid; grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));"
      />
    </div>
  </body>,
  "container": <div>
    <div
      style="display: grid; grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`List > renders the List component WHEN error 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Oops, something went wrong...
    </div>
  </body>,
  "container": <div>
    Oops, something went wrong...
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`List > renders the List component WHEN pending 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      Loading...
    </div>
  </body>,
  "container": <div>
    Loading...
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`List > renders the List component WITH values 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="display: grid; grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));"
      >
        <div
          style="width: 100px; height: 200px; margin: 0px auto;"
        >
          <div
            aria-description="abc"
            class="glass"
            style="width: 100%; height: 150px; background-position: center; background-repeat: no-repeat; background-size: contain; text-align: center; background-image: url("");"
          />
          <span
            style="font-size: 10px;"
          >
            abc
          </span>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="display: grid; grid-template-columns: repeat(auto-fit, minmax(100px, 1fr));"
    >
      <div
        style="width: 100px; height: 200px; margin: 0px auto;"
      >
        <div
          aria-description="abc"
          class="glass"
          style="width: 100%; height: 150px; background-position: center; background-repeat: no-repeat; background-size: contain; text-align: center; background-image: url("");"
        />
        <span
          style="font-size: 10px;"
        >
          abc
        </span>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
